### PR TITLE
Fix getting attributes for sharing violation files

### LINKED ISSUE #29782: Can no longer get file attributes from exclusively locked files
### Problem ###
IO.File.GetAttributes() doesn't return attributes for exclusively locked file like c:\pagefile.sys
IO.File.Exist() doesn't found such file (return false).

### Repo on PowerShell ###
```PowerShell
#########
[IO.File]::GetAttributes("c:\pagefile.sys")

Exception calling "GetAttributes" with "1" argument(s): "The process cannot access the file 'c:\pagefile.sys' because it is being used by another process."
At line:1 char:1
+ [IO.File]::GetAttributes("c:\pagefile.sys")
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
+ FullyQualifiedErrorId : IOException


$error[0].exception.InnerException | fl -Force

Message        : The process cannot access the file 'c:\pagefile.sys' because it is being used by another process.
Data           : {}
InnerException :
TargetSite     : System.IO.FileAttributes GetAttributes(System.String)
StackTrace     :    at System.IO.FileSystem.GetAttributes(String fullPath)
                    at CallSite.Target(Closure , CallSite , Type , String )
HelpLink       :
Source         : System.IO.FileSystem
HResult        : -2147024864


#########
[IO.File]::Exists("c:\pagefile.sys")

false
```

In PowerShell repo we have a workaround:
1. https://github.com/PowerShell/PowerShell/blob/5cbf1b43de52fe9025bf756990b1cc6eb4e694e9/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L6994-L7013
2. https://github.com/PowerShell/PowerShell/blob/5cbf1b43de52fe9025bf756990b1cc6eb4e694e9/src/System.Management.Automation/engine/Utils.cs#L989-L1029

Without the workaround PowerShell doesn't show the file:
```powershell
dir c:\pagefile.sys -Hidden
```
but cmd.exe show the file:
```
C:\Users\sie>dir c:\pagefile.sys /AH
 Volume in drive C has no label.
 Volume Serial Number is 824F-BBDF

 Directory of c:\

04/18/2018  10:48 AM    11,811,160,064 pagefile.sys
               1 File(s) 11,811,160,064 bytes
               0 Dir(s)  236,808,941,568 bytes free
```
### Suggestion ###
I believe the issue should be fixed in CoreFX here
https://github.com/dotnet/corefx/blob/f25eb288a449010574a6e95fe298f3ad880ada1e/src/System.IO.FileSystem/src/System/IO/FileSystem.Windows.cs#L203-L223

ERROR_SHARING_VIOLATION should be added.

### PR DESCRIPTION
**Pull request by JeremyKuhne**
Some files, such as the pagefile, refuse to allow you to get attributes directly.
You need to enumerate to get the attributes.

Enumerating is expensive, so we had changed from an opt-out to an opt-in. As
it took nearly a year to find this case, switching back to an opt-out with a
larger set of known cases where we get no value from trying again.

Also add an assert to flush out other error codes and scenarios.

Relates to #29782

Note that I couldn't repro the failure by manually locking our own file, but added cases nonetheless.

cc: @iSazonov

### CODE REVIEW

File: src/System.IO.FileSystem/tests/FileInfo/Exists.cs [hunk 1]
```diff
@@ -130,5 +130,54 @@ public void SymLinksMayExistIndependentlyOfTarget()
...
+
+        [Fact]
+        public void LockedFileExists()
+        {
+            string path = GetTestFilePath();
+            File.WriteAllBytes(path, new byte[10]);
+
+            using (FileStream stream = new FileStream(path, FileMode.Open, FileAccess.ReadWrite, FileShare.None))
+            {
+                stream.Lock(0, 10);
```
Comment (danmosemsft) [inline src/System.IO.FileSystem/tests/FileInfo/Exists.cs]: This is fine but I notice in the docs for ::LockFile that the length can be longer than the file so I guess Int32.MaxValue would work also..

File: src/System.IO.FileSystem/src/System/IO/FileSystem.Windows.cs [hunk 2]
```diff
@@ -200,12 +201,31 @@ internal static int FillAttributeInfo(string path, ref Interop.Kernel32.WIN32_FI
...
+                        && errorCode != Interop.Errors.ERROR_PATH_NOT_FOUND
+                        && errorCode != Interop.Errors.ERROR_NOT_READY
+                        && errorCode != Interop.Errors.ERROR_INVALID_NAME
+                        && errorCode != Interop.Errors.ERROR_BAD_PATHNAME
+                        && errorCode != Interop.Errors.ERROR_BAD_NETPATH
+                        && errorCode != Interop.Errors.ERROR_BAD_NET_NAME
+                        && errorCode != Interop.Errors.ERROR_INVALID_PARAMETER)
                     {
+                        // Assert so we can track down other cases (if any) to add to our test suite
+                        Debug.Assert(errorCode == Interop.Errors.ERROR_ACCESS_DENIED || errorCode == Interop.Errors.ERROR_SHARING_VIOLATION,
```
Comment (danmosemsft) [inline src/System.IO.FileSystem/src/System/IO/FileSystem.Windows.cs]: possibly ERROR_INVALID_PASSWORD ?

Comment (danmosemsft) [reply src/System.IO.FileSystem/src/System/IO/FileSystem.Windows.cs]: That's a total guess BTW I just noticed it when I looked up some of the others above.

File: src/System.IO.FileSystem/tests/FileInfo/Exists.cs [hunk 3]
```diff
@@ -130,5 +130,54 @@ public void SymLinksMayExistIndependentlyOfTarget()
...
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void PageFileExists()
+        {
+            // Typically there is a page file on the C: drive, if not, don't bother trying to track it down.
+            string pageFilePath = Directory.EnumerateFiles(@"C:\", "pagefile.sys").FirstOrDefault();
+            if (pageFilePath != null)
+            {
+                FileInfo info = new FileInfo(pageFilePath);
+                Assert.True(info.Exists);
```
Comment (pjanotti) [inline src/System.IO.FileSystem/tests/FileInfo/Exists.cs]: nit: perhaps coalesce this test with PageFileHasTime? While technically asserting different things it is one less test.

**Comment by danmosemsft** [comment]
Edited top comment to remove "fixes" as we need the issue open for servicing port

**Comment by JeremyKuhne** [comment]
Nano is failing for #29757;

**Comment by iSazonov** [comment]
@JeremyKuhne Many thanks for fast fix!

**Comment by JeremyKuhne** [comment]
Note that this didn't actually get reverted- the linked revert commit doesn't have any content.

**Comment by pjanotti** [comment]
Is/was there an issue that warranted revert of this PR?

**Comment by JeremyKuhne** [comment]
> Is/was there an issue that warranted revert of this PR?

No, this is apparently just some strange housekeeping by GitHub that popped up as a side effect of another reversion. The commit is still in the mainline.

**Comment by stephentoub** [comment]
No.  I clicked Revert on a completed unrelated PR, and for some reason GitHub decided to create 9 empty commits that were named with other PRs (note that the aforementioned commit is completely empty, zero diffs).  I abanonded that Revert as a result, but once on GitHub, forever on GitHub.

**Comment by iSazonov** [comment]
>  I clicked Revert on a completed unrelated PR and for some reason GitHub decided to create 9 empty commits

We saw this in PowerShell Core repo. Seems GitHub revert merged commit and then source commits. It wrong behavior. We decided that it was better to revert PR merged commit from command line.

**Pull request merged**